### ISSUE TITLE
fromAjax with automatic abort on becoming unobserved

### ISSUE THREAD

**Issue by jamiewinder** [mobxjs/mobx-utils#55]
I have a few scenarios like this where I have a `@computed` property which returns a `fromPromise` result:
```typescript
@observable filter = '';

@computed get filterItems() {
    const url = `/loaditems?filter=${this.filter}`;
    return fromPromise(getItemsFromUrl(url));
}
```
This works really well but if the observables that `filterItems` uses changes often, specifically mid-flight, we end up continuing a request that is no longer needed. I wonder if a special `fromAjax` utility might be useful here, which'd have the following behaviour:

- The result would have the same shape as `IPromiseBasedObservable`, however the state would have an additional value of 'cancelled'.
- Once created, if the result value of `fromAjax` become unobserved, the underlying AJAX request would be cancelled, and the state of the result would permanently transition to 'cancelled'.

This should mean that code like the snippet above should work well; recomputations which cause the `fromAjax` result to become unobserved will automatically abort the request in favour of a new request.

I wanted to float the idea before giving it a stab; please poke holes in anything I've posted above!

**Comment by urugator** [comment]
Maybe I don't understand, but whats do point of setting some state (`promise.cancelled`) or preventing a state update (`promise.value = response.result)` of the observable (`promise`), which is no longer observed (nothing is interested about that state)?
And if something later re-subscribes for that observable it means that the state is again relevant, therefore should return the resolved (or not yet resolved) value.

**Comment by jamiewinder** [comment]
That part of this is just for completeness. If you have access to an `IPromiseBasedObservable` and its state is 'cancelled' then you wouldn't expect it to ever have a value. 

The key part of this suggestion is that the underlying AJAX call would be aborted if it's still pending and there are no observers of it. A bit like a `switchMap` / `flatMapLatest` in RxJS.

**Comment by urugator** [comment]
> That part of this is just for completeness. If you have access to an IPromiseBasedObservable and its state is 'cancelled' then you wouldn't expect it to ever have a value.

I could understand this only if you would have to manually cancel the promise `promise.cancel()`, but it doesn't make sense to me to cancel the promise by becoming unobserved.

> AJAX call would be aborted

ajax call is made immeditiately when `filterItems` is called, you can only abort update of `promis.value` when response arrives...

**Comment by jamiewinder** [comment]
If we imagine a component using [the return value of] `filterItems`, when would it become unobserved? When the filter changes, or when the component is unmounted. In both cases, you don't need the AJAX call to continue, so it might as well be aborted. For example, if your loaditems service can take a few seconds to complete, then you don't want your browser continuing to wait for the results (or the service continuing to service the request) if your application is no longer waiting on them. It seems to be that you can detect this in the MobX world in terms of 'is anyone still observing this?'. If yes, continue. If not, abort.

There might be some circumstances where you don't want this behaviour, but in those you could revert to using `fromPromise`.

**Comment by urugator** [comment]
> When the filter changes, or when the component is unmounted. 

is not always the same as

>  'is anyone still observing this?'.

and in cases where it is the same, it would mean that modifing the status of that promise in any way (setting cancelled or whatever) is not relevant, because **nothing is observing** ... yes something can still read the promise outside of reaction (observer), but you would make the behavior inconsistent based on how do you consume the promise...which is just wrong... (if you would use reaction, the promise woudn't be cancelled, because there is still this one reaction, even if component is unmounted...)

I think it makes sense to make promise manually cancellable and provide something like `onBecomeUnobserved` as a MobX alternative to a react specific `componentWillUnmount`, so you can have I fine-grained control like:
`onBecomeUnobserved(promise, () => promise.cancel())`
Othrewise it's just mess imho...

**Comment by mweststrate** [comment]
Afaik cancellable promises are not standardized yet, so it would be hard to provide a generic solution to this problem? (Or is my knowledge outdated :-P) Probably it is best to set up a request in your computed prop yourself, cancelling any current ones if the computed value is invoked again?

**Comment by urugator** [comment]
I think the idea is to simply cancel the update of the observables when actual promise is resolved/rejected. Something like:

```javascript
// from-promise.ts
constructor(public promise: PromiseLike < T >, initialValue: T = undefined) {
  this._observable = observable.box(initialValue);
  promise.then(
    action("observableFromPromise-resolve", (value: T) => {
      if (this._state !== CANCELLED) { // <-- 
        this._observable.set(value);
        this._state.set("fulfilled");
      }
    }),
    action("observableFromPromise-reject", (reason: any) => {
      if (this._state !== CANCELLED) { // <--
        this._reason.set(reason);
        this._observable.set(reason);
        this._state.set("rejected");
      }
    })
  );
}  
```
Guess it may be handy when there is some long running operation with a cancel button ...

**Comment by jamiewinder** [comment]
Cancellable promises aren't a thing, but `abort`-able AJAX calls are, which is why I went for the non-generic suggestion and name. I think the `onBecomeUnobserved` utility like @urugator suggested should make this simple to implement; basically the same as `fromPromise`, but with:
```
onBecomeUnobserved(() => {
   xhr.abort();
   fromAjaxResult.state = 'aborted';
});
```

**Comment by mweststrate** [comment]
MobX 4 exposes an `onbecomeUnobserved` hooks, which can be used to abort ajax requests like suggested above, and as demonstrated in https://codesandbox.io/s/l73yprwj9z.
